java/flight/flight-jdbc-driver/src/main/java/org/apache/arrow/driver/jdbc/accessor/impl/calendar/ArrowFlightJdbcDateVectorAccessor.java

```diff
@@ -94,7 +97,23 @@ public Date getDate(Calendar calendar) {
     long value = holder.value;
     long milliseconds = this.timeUnit.toMillis(value);
 
-    return new Date(DateTimeUtils.applyCalendarOffset(milliseconds, calendar));
+    long millisWithCalendar = DateTimeUtils.applyCalendarOffset(milliseconds, calendar);
+
+    return new Date(getTimestampValue(millisWithCalendar).getTime());
+  }
+
+  private Timestamp getTimestampValue(long millisWithCalendar) {
+    long milliseconds = millisWithCalendar;
+    if (milliseconds < 0) {
+      // LocalTime#ofNanoDay only accepts positive values
+      milliseconds -= ((milliseconds / MILLIS_PER_DAY) - 1) * MILLIS_PER_DAY;
+    }
+
+    return Timestamp.valueOf(
+        LocalDateTime.of(
+            LocalDate.ofEpochDay(millisWithCalendar / MILLIS_PER_DAY),
+            LocalTime.ofNanoOfDay(TimeUnit.MILLISECONDS.toNanos(milliseconds % MILLIS_PER_DAY)))
+    );
```

Comment:
Why not just use new Timestamp()?

Reply:
I've test this method with a negative value for the miliseconds and it is giviing me an incorrect date

Reply:
Creating Timestamp with millis is the reason for failure. The only way of getting accurate values for the Date part is using LocalDateTime because it handles negative timestamps better, and the constructor benefits from it by using specific getter methods in it to create itself.

Reply:
Moved to utils, and created unit tests for negative scenarios.